File: 0001-0500/0022/0022.generate-parentheses.cs
/*
 * @lc app=leetcode id=22 lang=csharp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var result = new List<string>();
        GenerateParenthesisHelper(result, "", 0, 0, n);
        return result;
    }

    public void GenerateParenthesisHelper(List<string> result, string current, int open, int close, int max) {
        if(current.Length == max * 2) {
            result.Add(current);
            return;
        }
        if(open < max) {
            GenerateParenthesisHelper(result, current + "(", open + 1, close, max);
        }
        if(close < open) {
            GenerateParenthesisHelper(result, current + ")", open, close + 1, max);
        }
    }
}
// @lc code=end

File: 0001-0500/0022/0022.generate-parentheses.js
/*
 * @lc app=leetcode id=22 lang=javascript
 *
 * [22] Generate Parentheses
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    let result = [];
    let generate = (left, right, s) => {
        if (left === n && right === n) {
            result.push(s);
            return;
        }
        if (left < n) generate(left + 1, right, s + '(');
        if (right < left) generate(left, right + 1, s + ')');
    };
    generate(0, 0, '');
    return result;
};
// @lc code=end

